### PR TITLE
A lot of stuff

### PR DESCRIPTION
**Pull request by well-it-wasnt-me**
## Description
Cleaned even more the code, fixed some logical error and incoherence in the code, going crazy with test cases and mock, readme update, first doc draf

## Changes Made
- sas.py = Many function that should have returned a list of stuff would return just a single byte. Fixed, now a list is returned. Reformatted with black, use of models
- models/ = Create classes for everything needed with a proper method
- tests/ = Still a skeleton, trying to understand why pytest cant see pythoncrc and kermit

## Additional Notes
I am still without a gaming machine to try out the code....so...again...test before merging in master

### CODE REVIEW

**Comment by dedalgr** [comment]
Many machines do not return anything to the command and the only thing you
can check if the cable is good and you still have a connection is by
reading 1 bit.

На нд, 28.01.2024 г. в 16:06 Antonio ***@***.***> написа:

**Comment by well-it-wasnt-me** [comment]
> Many machines do not return anything to the command and the only thing you can check if the cable is good and you still have a connection is by reading 1 bit. 

Yes, i realized it reading the SAS documentation (6.02). In fact i've implemented "complex" responses where needed. In other part i left the 1 bit. Planning to make those (with 1 bit) as a simple bool function

**Comment by dedalgr** [comment]
You don't need complicated answers. Use events poll but predict generation
time

На нд, 28.01.2024 г. в 16:17 Antonio ***@***.***> написа:

**Comment by well-it-wasnt-me** [comment]
my bad. by "complex" i meant show the list of key => value pair like showing
`
["Gaming machine registered": 01 ]
`

i am not so sure by what you mean with 

> Use events poll but predict generation time

**Comment by dedalgr** [comment]
data = sas.initiate_legacy_bonus_pay(10, tax=1, games=1)
if data is True:
while 1:
if sas.events_poll() == 'A legacy bonus pay awarded and/or a multiplied
jackpot occurred':
break

На нд, 28.01.2024 г. в 16:21 Григор ***@***.***> написа:

**Comment by dedalgr** [comment]
Use events poll but predict generation time

На нд, 28.01.2024 г. в 16:27 Григор ***@***.***> написа:

**Comment by well-it-wasnt-me** [comment]
i learned that `event_poll` take the events from a FIFO memory stacking messages. So, when calling `event_poll` you do not get the actual event but the first entered in it's memory. For this reason i dont think is very reliable.

And use a string like that is not very coherent with actual developing standards, thats why i've attached it to the model class: better and faster recognition and if in SAS 6.04 they change some reply is easier to update

**Comment by dedalgr** [comment]
I can show you via anydesk how the whole protocol is used and how it works
100% on a real machine

На нд, 28.01.2024 г. в 16:27 Григор ***@***.***> написа:

**Comment by well-it-wasnt-me** [comment]
i've been working on 3 different real machines till few days ago (waiting they fix the vpn in the office).

I'm sure that you are right 100% and that your code is bulletproof...but i also realized that many manufacturer implement the SAS Protocol a bit like "as they wish". So, not every machine is coherent in their reply,

So, my intent, is to provide a library with the outmost compatibility (with protocol and machines), readability and easyness possible

**Comment by dedalgr** [comment]
You got it right, manufacturers implement the protocol differently. I have
tested it on EGT, CT, Amatic, Novomatic, Mercur and IGT. Like most, I have
tried them with different computers because Casino Technology with ARM
processor does not work like with Intel. Amatic with Windows works
differently than Amatic with Linux

На нд, 28.01.2024 г. в 16:39 Antonio ***@***.***> написа:

**Comment by dedalgr** [comment]
Anyway, after 4 days I disappear from github and the communication ends. Do
whatever you decide. I on the other hand use it in real casinos with almost
all machine manufacturers and versions and sell software. This is how I pay
my bills. I wanted to help and you have 4 days after which I won't fulfill
the github login requirement with a phone and I won't respond again


На нд, 28.01.2024 г. в 17:06 Григор ***@***.***> написа:

**Comment by well-it-wasnt-me** [comment]
Thank @dedalgr !!!! We will for sure be in contact !

**Comment by dedalgr** [comment]
You can not connect after 4 days or not for free. You have 4 days to
understand my code modification. After 4 days, the story ends. Ask smart
questions very quickly

На нд, 28.01.2024 г. в 17:15 Antonio ***@***.***> написа:

**Comment by well-it-wasnt-me** [comment]
Yes, i figured :) I will for sure in the next hours

**Comment by dedalgr** [comment]
event_poll is absolutely unreliable unless you're looking for "AFT Host
Pull Request" or "Payout Button Pressed" with lots of conditions
With initiate_legacy_bonus_pay it is better to wait for a change in the bet
counter and send when the reel is spinning, then get 1 bit with the machine
number to see if you have a connection.

На нд, 28.01.2024 г. в 16:32 Antonio ***@***.***> написа:

**Comment by dedalgr** [comment]
Of course you can use event_poll but the time is not 1,2,3 seconds but it
jumps to 2 minutes

На нд, 28.01.2024 г. в 17:48 Григор ***@***.***> написа:

File: README.md [hunk 1]
```diff
@@ -37,10 +37,12 @@ sudo echo reboot
 connection:
   serial_port: /dev/serial0
 ```
-#### Connecting with RPI 4:
 
-## NOTE: To avoid frying your Pi, you are still going to need a rs-232 to TTL logic level converter. I used the [MAX3323](https://www.mouser.com/ProductDetail/Analog-Devices-Maxim-Integrated/MAX3323EEPE%2b?qs=CDqwynd4ZNoRwc1iI5RFww%3D%3D)
-#### Connecting with USB to serial adapter, use a usb rs232 cable.....I used this [one](https://www.amazon.com/USB-Serial-Adapter-Prolific-PL-2303/dp/B00GRP8EZU/ref=sr_1_1_sspa?dib=eyJ2IjoiMSJ9.eT7IwLbFTyi5P6wiZqvnXrIsQpdtfPz_M46xtQa_S1I6h-lpFonAvq5YC5xJqm4vO8e3APmv6ZveRIHnEk3JvZ7RPORl8CFQWSUM226Dz0JssJAFQzWxU_Rk-YZaVXY5yPT9ZX-bqG0CDKUEzPruTJWEFg-ITUZtUOwr8KLTrvxvVg-ounmiZNAaizmQvxjrTdVozOF4iRbI5UF54oqfyn1obbD9whyaS_eGnl-TRcU.CRPZSqj6-D9E9pUJExtcBxGZd89oO6OAewGmvDxATTU&dib_tag=se&keywords=prolific%2Busb%2Bto%2Bserial&qid=1705598420&sr=8-1-spons&sp_csd=d2lkZ2V0TmFtZT1zcF9hdGY&th=1).
+5) **NOTE** To avoid frying your Pi, you are still going to need a rs-232 to TTL logic level converter. I used the [MAX3323](https://www.mouser.com/ProductDetail/Analog-Devices-Maxim-Integrated/MAX3323EEPE%2b?qs=CDqwynd4ZNoRwc1iI5RFww%3D%3D)
+
+### Alternative to MAX3323 and Pi UART
```
Comment (zacharytomlinson) [inline README.md:43]: Just a clarification here, if using a usb to serial adapter you are still using a rs232 to ttl converter, it is just inside of your usb dongle. I bet if you cracked the dongle open it has a MAX, FTDI or similar clone chip in it. One method just accesses the UART via a usb port and the other uses the GPIO pins.

**Review by zacharytomlinson — APPROVED** [review]

**Comment by zacharytomlinson** [comment]
@dedalgr I have one question, I have access to most of the machines that run in the West so will have no problems doing testing and debugging there, but I am not familiar with what machines are common in Eastern Europe. I think you have stated a majority of the casinos your software interfaces with are in that area, so do you have a breakdown of what type of machines are most  common in that region?

**Comment by well-it-wasnt-me** [comment]
Im about to merge and close this pr. 
Please @dedalgr and @zacharytomlinson lets move this into the discussion tab. 
In the next days I'll prepare a wiki containing also what has been said here.

**Comment by dedalgr** [comment]
https://www.egt.com/bg/
https://ctgaming.com/
https://www.novomatic.com/en
https://www.amatic.com/products/

На нд, 28.01.2024 г. в 21:06 Zachary Tomlinson ***@***.***>
написа:

**Pull request merged**